diff --git a/typings/index.d.ts b/typings/index.d.ts
index bcbc6ff3faca553b8318964de52da9ab610f8330..80fb903e426a54551fcc382b7240012f85ed7410 100644
--- a/typings/index.d.ts
+++ b/typings/index.d.ts
@@ -49,7 +49,7 @@ declare module 'react-draggable' {
     offsetParent: HTMLElement,
     grid: [number, number],
     handle: string,
-    nodeRef?: React.RefObject<HTMLElement>,
+    nodeRef?: React.RefObject<HTMLElement | null>,
     onStart: DraggableEventHandler,
     onDrag: DraggableEventHandler,
     onStop: DraggableEventHandler,
